### DISCUSSION TITLE
mdx-js and TypeScript: What are the blockers from it from being rewritten in TS?

### DISCUSSION THREAD

**Discussion by tony** [mdx-js/mdx#1540] [Ideas] [answered]
Where does TypeScript stand with mdx-js 2.0 and beyond in the here and now?

The last mention is #1041, for 2.0, @ChristianMurphy introduced typings in #1078/#1079. This is exciting news!

What's keeping an even bigger change, e.g. full rewrite in TS from happening? What effort would that actually entail? The complexity of the types involving a lot of generics potentially?

This would be invaluable since it offers extra safety/completions when extending MDX.   From experience I highly prefer pure TS to generated or handmade typings. 

Do maintainers or contributors have enough spare time to rewrite in TypeScript if they were funded? If so is there a ballpark of how much they'd want raised?

**Comment by ChristianMurphy** [comment, marked as answer]
Unified, the underlying library supporting remark and mdx is in the process of adding native typings through TypeScript+JSDoc (see https://github.com/unifiedjs/rfcs/pull/5, https://github.com/unifiedjs/unified/issues/121, and https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html for more info).
Financial support to fund the effort is welcome! (https://opencollective.com/unified)

There may be interest in bringing it to MDX 2 as well.
Though disagreements around some of the planned features for version 2 have stalled new features and the GA release for a while.
Hopefully it can make it's way into MDX. In the meantime XDM (https://github.com/wooorm/xdm) offers another option, with native TS+JSDoc typings already provided.

**Reply by clicktodev** [reply]
Looks like the unified issues you've mentioned have been addressed.
do you think the repo is ready for TypeScript adoption now?

**Reply by ChristianMurphy** [reply]
@hichemfantar MDX is written with TypeScript using the JSDoc syntax already, is has been since MDX v2 (MDX is on v3 now and has even deeper TS integration).
There is no need to adopt what is already adopted 🙂

**Reply by clicktodev** [reply]
I meant writing actual typescript .ts files instead of JSDoc.
The advantages of typescript are much more than just type declarations.
The advantages are mainly for the project itself as typescript helps catch many bugs.

**Reply by ChristianMurphy** [reply]
@hichemfantar I'd recommend reading with JSDoc mode for TypeScript is https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html
All of the files are checked with TSC
https://github.com/mdx-js/mdx/blob/2b5ff6954475be3771c679f957f7cb49ded1df97/package.json#L132

With script compiler settings https://github.com/mdx-js/mdx/blob/2b5ff6954475be3771c679f957f7cb49ded1df97/tsconfig.json#L15

And every variable in the code has a type https://github.com/mdx-js/mdx/blob/2b5ff6954475be3771c679f957f7cb49ded1df97/package.json#L177

Converting to .ts adds an unnecessary build step, adds indirection, and adds nothing to the type safety (it already is safe).

**Reply by clicktodev** [reply]
Thanks for the explanation, I wasn't aware of those things so thanks again for the clarification!
